src/input/mod.rs

```diff
@@ -2171,7 +2172,12 @@ impl State {
             AxisSource::Finger => self.niri.config.borrow().input.touchpad.scroll_factor,
             _ => None,
         };
-        let scroll_factor = scroll_factor.map(|x| x.0).unwrap_or(1.);
+        let window_scroll_factor = self
+            .niri
+            .window_under_cursor()
+            .and_then(|window| window.rules().scroll_factor);
```

Comment:
So this probably works in most cases, but ideally you should call `update_pointer_contents()` from below first, then get `pointer.current_focus()` and then find the window for that WlSurface. This is because pointer grabs will modify the pointer focus under the hood.